From: Zi Yan <ziy@nvidia.com>
Subject: mm/memcg: use order instead of nr in split_page_memcg()
Date: Mon, 3 Apr 2023 16:18:33 -0400

Patch series "Split a folio to any lower order folios", v3.

File folio supports any order and people would like to support flexible
orders for anonymous folio[1] too.  Currently, split_huge_page() only
splits a huge page to order-0 pages, but splitting to orders higher than 0
is also useful.  This patchset adds support for splitting a huge page to
any lower order pages and uses it during file folio truncate operations.


This patch (of 7):

We do not have non power of two pages, using nr is error prone if nr is
not power-of-two.  Use page order instead.

Link: https://lkml.kernel.org/r/20230403201839.4097845-1-zi.yan@sent.com
Link: https://lkml.kernel.org/r/20230403201839.4097845-2-zi.yan@sent.com
Signed-off-by: Zi Yan <ziy@nvidia.com>
Cc: Kirill A. Shutemov <kirill.shutemov@linux.intel.com>
Cc: Matthew Wilcox (Oracle) <willy@infradead.org>
Cc: Michal Koutný <mkoutny@suse.com>
Cc: Roman Gushchin <roman.gushchin@linux.dev>
Cc: Ryan Roberts <ryan.roberts@arm.com>
Cc: Yang Shi <shy828301@gmail.com>
Cc: Yu Zhao <yuzhao@google.com>
Cc: Zach O'Keefe <zokeefe@google.com>
Signed-off-by: Andrew Morton <akpm@linux-foundation.org>
---

 include/linux/memcontrol.h |    4 ++--
 mm/huge_memory.c           |    3 ++-
 mm/memcontrol.c            |    3 ++-
 mm/page_alloc.c            |    4 ++--
 4 files changed, 8 insertions(+), 6 deletions(-)

--- a/include/linux/memcontrol.h~mm-memcg-use-order-instead-of-nr-in-split_page_memcg
+++ a/include/linux/memcontrol.h
@@ -1152,7 +1152,7 @@ static inline void memcg_memory_event_mm
 	rcu_read_unlock();
 }
 
-void split_page_memcg(struct page *head, unsigned int nr);
+void split_page_memcg(struct page *head, int order);
 
 unsigned long mem_cgroup_soft_limit_reclaim(pg_data_t *pgdat, int order,
 						gfp_t gfp_mask,
@@ -1593,7 +1593,7 @@ void count_memcg_event_mm(struct mm_stru
 {
 }
 
-static inline void split_page_memcg(struct page *head, unsigned int nr)
+static inline void split_page_memcg(struct page *head, int order)
 {
 }
 
--- a/mm/huge_memory.c~mm-memcg-use-order-instead-of-nr-in-split_page_memcg
+++ a/mm/huge_memory.c
@@ -2512,10 +2512,11 @@ static void __split_huge_page(struct pag
 	struct address_space *swap_cache = NULL;
 	unsigned long offset = 0;
 	unsigned int nr = thp_nr_pages(head);
+	int order = folio_order(folio);
 	int i;
 
 	/* complete memcg works before add pages to LRU */
-	split_page_memcg(head, nr);
+	split_page_memcg(head, order);
 
 	if (PageAnon(head) && PageSwapCache(head)) {
 		swp_entry_t entry = { .val = page_private(head) };
--- a/mm/memcontrol.c~mm-memcg-use-order-instead-of-nr-in-split_page_memcg
+++ a/mm/memcontrol.c
@@ -3441,11 +3441,12 @@ void obj_cgroup_uncharge(struct obj_cgro
 /*
  * Because page_memcg(head) is not set on tails, set it now.
  */
-void split_page_memcg(struct page *head, unsigned int nr)
+void split_page_memcg(struct page *head, int order)
 {
 	struct folio *folio = page_folio(head);
 	struct mem_cgroup *memcg = folio_memcg(folio);
 	int i;
+	unsigned int nr = 1 << order;
 
 	if (mem_cgroup_disabled() || !memcg)
 		return;
--- a/mm/page_alloc.c~mm-memcg-use-order-instead-of-nr-in-split_page_memcg
+++ a/mm/page_alloc.c
@@ -2781,7 +2781,7 @@ void split_page(struct page *page, unsig
 	for (i = 1; i < (1 << order); i++)
 		set_page_refcounted(page + i);
 	split_page_owner(page, 1 << order);
-	split_page_memcg(page, 1 << order);
+	split_page_memcg(page, order);
 }
 EXPORT_SYMBOL_GPL(split_page);
 
@@ -4997,7 +4997,7 @@ static void *make_alloc_exact(unsigned l
 		struct page *last = page + nr;
 
 		split_page_owner(page, 1 << order);
-		split_page_memcg(page, 1 << order);
+		split_page_memcg(page, order);
 		while (page < --last)
 			set_page_refcounted(last);
 
_
